New sections for lexical structure, ddl, and primer docs (#6041)
confluentinc/ksql: docs/reference/sql/data-definition.md

keywords: ksqldb, sql, ddl
---

This section covers how you create the structures that store your events.
ksqlDB abstracts events as rows with columns and stores them in streams
and tables.

## Rows and columns

Streams and tables help you model collections of events that accrete over time.
Both are represented as a series of rows and columns with a schema, much like a
relational database table. Rows represent individual events. Columns represent
the attributes of those events.

Each column has a data type. The data type limits the span of permissible values
that you can assign. For example, if a column is declared as type `INT`, it can't
be assigned the value of string `'foo'`.

In contrast to relational database tables, the columns of a row in ksqlDB are
divided into _key_ and _value_ columns. The key columns control which partition
a row resides in. The value columns, by convention, store the main data of
interest. Controlling the key columns is useful for manipulating the underlying
data locality, and enables you to integrate with the wider {{ site.ak }}
ecosystem, which uses the same key/value data model. By default, a column is a

> value column. Marking a column as a `(PRIMARY) KEY` makes it a key column.

Reviewer: May be worth also mentioning that they key column also controls ordering guarantees, i.e. only rows in the same partition have any relative ordering guarantees.

Internally, each row is backed by a [Kafka record](../../../overview/apache-kafka-primer/#records).
In {{ site.ak }}, the key and value parts of a record are
[serialized](../../../overview/apache-kafka-primer/#serializers) independently.
ksqlDB enables you to exercise this same flexibility and builds on the semantics
of {{ site.ak }} records, rather than hiding them.

There is no theoretical limit on the number of columns in a stream or table.
In practice, the limit is determined by the maximum message size that {{ site.ak }}

> can store and the resources dedicated to ksqlDB.

Reviewer: Worth mentioning that large schemas generally mean large serde costs?   Good to encourage people to keep their schemas simple!

## Streams

A stream is a partitioned, immutable, append-only collection that represents a
series of historical facts. For example, the rows of a stream could model a
sequence of financial transactions, like "Alice sent $100 to Bob", followed by
"Charlie sent $50 to Bob".

Once a row is inserted into a stream, it can never change. New rows can be
appended at the end of the stream, but existing rows can never be updated or

> deleted.

Reviewer: worth mentioning that while rows can be deleted, they can be 'aged out' and link to the retention bit above?

Each row is stored in a particular partition. Every row, implicitly or explicitly,
has a key that represents its identity. All rows with the same key reside in the

> same partition.

Reviewer: > Every row, implicitly or explicitly, has a key that represents its identity.

This isn't true - ksqlDB supports keyless streams.  Keyless streams ignore the key of the source and produce to the output with no key set (meaning rows are randomly assigned a partition).

To create a stream, use the `CREATE STREAM` command. The following example
statement specifies a name for the new stream, the names of the columns, and
the data type of each column.

```sql
CREATE STREAM s1 (
    k VARCHAR KEY,
    v1 INT,
    v2 VARCHAR
) WITH (
    kafka_topic = 's1',
    partitions = 3,
    value_format = 'json'
);
```

This creates a new stream named `s1` with three columns: `k`, `v1`, and `v2`.
The column `k` is designated as the key of this stream, which controls the
partition that each row is stored in. When the data is stored, the value
portion of each row's underlying {{ site.ak }} record is serialized in the
JSON format.

Under the hood, each stream corresponds to a [Kafka topic](../../../overview/apache-kafka-primer/#topics)

> with a registered schema. If the backing topic for a stream doesn't exist when

Reviewer: ```suggestion
with an associated schema. If the backing topic for a stream doesn't exist when
```

you declare it, ksqlDB creates it on your behalf, as shown in the previous
example statement.

> You can also declare a stream on top of an existing topic. When you do that,
> ksqlDB simply registers its associated schema. If topic `s2` already exists,
> the following statement register a new stream over it:

Reviewer: ```suggestion
You can also declare a stream on top of an existing topic. When you do that,
ksqlDB simply associates the schema with the topic. If topic `s2` already exists,
the following statement register a new stream over it:
```

```sql
CREATE STREAM s2 (
    k1 VARCHAR KEY,
    v1 VARCHAR
) WITH (
    kafka_topic = 's2',
    value_format = 'json'
);
```

!!! tip
    When you create a stream on an existing topic, you don't need to declare
    the number of partitions for the topic. ksqlDB infers the partition count
    from the existing topic.

## Tables

A table is a mutable, partitioned collection that models change over time. In
contrast with a stream, which represents a historical sequence of events, a
table represents what is true as of "now". For example, you might use a table
to model the locations where someone has lived as a stream: first Miami, then

> New York, then London, and so forth.

Reviewer: >  a table represents what is true as of "now"

Hummm... you can also use a table to ask 'what was true as of <some time>'.   So I'm not sure this is a good description of a table.

> Tables work by leveraging the keys of each row. If a sequence of rows shares a
> key, the last row for a given key represents the most up-to-date information
> for that key's identity. A background process periodically runs and deletes all
> but the newest rows for each key.

Reviewer: I think we should be careful in our use of the terms `row` and `record` or `event`.

In my mind, `row` is referring to the row in the table, where are `record` or `event` is the record coming from the Kafka topic we're processing, that may mutate the row in the table.   

A non-temporal table only has one row with a specific key. Hence, IMHO, the phrase "If a sequence of rows share a key" is unintuitive ... as only one row can have a specific key.

```suggestion
Tables work by leveraging the keys of each row. If a sequence of records shares a
key, the last record for a given key represents the most up-to-date row
for that key's identity. A background process periodically runs and deletes all
but the newest record for each key.
```

Reviewer: Maybe add link to compaction in the last sentence?

example statement declares a `current_location` table that has a key field 
named `person`.

```sql
CREATE TABLE current_location (
    person VARCHAR PRIMARY KEY,
    location VARCHAR
) WITH (
    kafka_topic = 'current_location',
    partitions = 3,
    value_format = 'json'
);
```

As with a stream, you can declare a table directly on top of an existing
{{ site.ak }} topic by omitting the number of partitions in the `WITH` clause.

## Keys

You can mark a column with the `KEY` keyword to indicate that it's a key
column. Key columns constitute the key portion of the row's underlying
{{ site.ak }} record. Only streams can mark columns as keys, and it's optional
for them to do do. Tables must use the `PRIMARY KEY` constraint instead.

In the following example statement, `k1`'s data is stored in the key portion of
the row, and `v1`'s data is stored in the value.

```sql
CREATE STREAM s3 (
    k1 VARCHAR KEY,
    v1 VARCHAR
) WITH (
    kafka_topic = 's3',
    value_format = 'json'
);
```

The ability to declare key columns explicitly is especially useful when you're
creating a stream over an existing topic. If ksqlDB can't infer what data is in

> the key of the underlying {{ site.ak }} record, it must perform a repartition

Reviewer: > If ksqlDB can't infer what data is in the key of the underlying {{ site.ak }} record...

I find this misleading. ksqlDB never tries to infer what's in the key of the record.  The user must _tell_ ksqlDB what the key is...

> of the rows internally. If you're not sure what data is in the key or you simply
> don't need it, you can omit the `KEY` keyword.

Reviewer: > If you're not sure what data is in the key or you simply don't need it, you can omit the `KEY` keyword.

If find this misleading.  This suggests omitting just the `KEY` keyword is sufficient for ksqlDB to load the `k1` column in the above example correctly.  

```suggestion
of the rows internally. If you're not sure what data is in the key or you simply
don't need it, you can omit the `KEY` column. For example:

```sql
CREATE STREAM s3 (
    v1 VARCHAR
) WITH (
    kafka_topic = 's3',
    value_format = 'json'
);
```

## Default values

If a column is declared in a schema, but no attribute is present in the
underlying {{ site.ak }} record, the value for the row's column is populated as
`null`.

## Pseudocolumns

A pseudocolumn is a column that's automatically populated by ksqlDB and contains
meta-information that can be inferred about the row at creation time. By default,
pseudocolumns aren't returned when selecting all columns with the star (`*`)
special character. You must select them explicitly, as shown in the following
example statement.

```sql
SELECT ROWTIME, * FROM s1 EMIT CHANGES;
```

> The following table lists all pseudocolumns.

Reviewer: They're currently not defined as pseudo columns. They are included by default with a `select *`.

I guess at the moment they're defined a 'system columns'. There names are reserved for system use.

| pseudocolumn | meaning                        |
|--------------|--------------------------------|
| `ROWTIME`    | Row timestamp, inferred from the underlying Kafka record if not overridden. |

You can't create additional pseudocolumns beyond these.

## Constraints

Although data types help limit the range of values that can be accepted by
ksqlDB, sometimes it's useful to have more sophisticated restrictions.
_Constraints_ enable you to exercise this type of logic directly in your schema.

### Primary key constraints

In a relational database, a primary key indicates that a column will be used as
a unique identifier for all rows in a table. If you have a table that has a row
with primary key `5`, you can't insert another row whose primary key is also `5`.

ksqlDB uses primary keys in a similar way, but there are a few differences,
because ksqlDB is an event streaming database, not a relational database.

- Only tables can have primary keys. Streams do not support them.
- Adding multiple rows to a table with the same primary key doesn't cause the

>   subsequent rows to be rejected.

Reviewer: This is mixing up the concept of a `table` with a `changelog`.

The table can only have one row with a specific primary key.  A changelog can contain multiple rows with the same key.  The changelog can be materialized into a table.

This is the same as a traditional database.

We should explain this table/changelog duality somewhere ;)

Yes, unfortunately at the moment we can `INSERT VALUES` into a table with an key for a row that already exists and it will work.  However, this is a bug IMHO, (it should be `UPSERT` not `INSERT`),.
